### PR TITLE
[VarDumper] Support for ReflectionAttribute

### PR DESCRIPTION
**Pull request by derrabus**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | N/A 
| License       | MIT
| Doc PR        | not needed

VarDumper currently does not understand that certain reflection objects might have attributes attached to it. Dumping a `ReflectionAttribute` just yields `ReflectionAttribute {#4711}` which is not really helpful. This PR attempts to fix this.

```
ReflectionAttribute {#4711
  name: "App\MyAttribute"
  arguments: array:2 [
    0 => "one"
    "extra" => "hello"
  ]
}
```

While working on this, I noticed that class constants (which can be reflected on since PHP 7.1) are just dumped as plain values, so I've also added a caster for `ReflectionClasConstant` as bonus.

The full output for the `LotsOfAttributes` fixture class that is included with is PR looks like this:

<details>

```
^ ReflectionClass {#7
  +name: "Symfony\Component\VarDumper\Tests\Fixtures\LotsOfAttributes"
  modifiers: "final"
  implements: []
  constants: array:1 [
    0 => ReflectionClassConstant {#20
      +name: "SOME_CONSTANT"
      +class: "Symfony\Component\VarDumper\Tests\Fixtures\LotsOfAttributes"
      modifiers: "public"
      value: "some value"
      attributes: array:2 [
        0 => ReflectionAttribute {#33
          name: "Symfony\Component\VarDumper\Tests\Fixtures\RepeatableAttribute"
          arguments: array:1 [
            0 => "one"
          ]
        }
        1 => ReflectionAttribute {#34
          name: "Symfony\Component\VarDumper\Tests\Fixtures\RepeatableAttribute"
          arguments: array:1 [
            0 => "two"
          ]
        }
      ]
    }
  ]
  properties: array:1 [
    "someProperty" => ReflectionProperty {#19
      +name: "someProperty"
      +class: "Symfony\Component\VarDumper\Tests\Fixtures\LotsOfAttributes"
      modifiers: "private"
      attributes: array:1 [
        0 => ReflectionAttribute {#30
          name: "Symfony\Component\VarDumper\Tests\Fixtures\MyAttribute"
          arguments: array:2 [
            0 => "one"
            "extra" => "hello"
          ]
        }
      ]
    }
  ]
  methods: array:1 [
    "someMethod" => ReflectionMethod {#21
      +name: "someMethod"
      +class: "Symfony\Component\VarDumper\Tests\Fixtures\LotsOfAttributes"
      returnType: "void"
      parameters: {
        $someParameter: ReflectionParameter {#28
          +name: "someParameter"
          position: 0
          attributes: array:1 [
            0 => ReflectionAttribute {#42
              name: "Symfony\Component\VarDumper\Tests\Fixtures\MyAttribute"
              arguments: array:1 [
                0 => "three"
              ]
            }
          ]
          typeHint: "string"
        }
      }
      attributes: array:1 [
        0 => ReflectionAttribute {#27
          name: "Symfony\Component\VarDumper\Tests\Fixtures\MyAttribute"
          arguments: array:1 [
            0 => "two"
          ]
        }
      ]
      modifiers: "public"
    }
  ]
  attributes: array:1 [
    0 => ReflectionAttribute {#22
      name: "Symfony\Component\VarDumper\Tests\Fixtures\MyAttribute"
      arguments: []
    }
  ]
  extra: {
    file: "./src/Symfony/Component/VarDumper/Tests/Fixtures/LotsOfAttributes.php"
    line: "15 to 28"
    isUserDefined: true
  }
}
```

</details>

### CODE REVIEW

File: src/Symfony/Component/VarDumper/Caster/ReflectionCaster.php [hunk 1]
```diff
@@ -330,7 +358,10 @@ private static function addExtra(&$a, \Reflector $c)
         }
     }
 
-    private static function addMap(&$a, \Reflector $c, $map, $prefix = Caster::PREFIX_VIRTUAL)
+    /**
+     * @param \Reflector|\ReflectionAttribute $c
+     */
+    private static function addMap(array &$a, $c, $map, $prefix = Caster::PREFIX_VIRTUAL)
```
Comment (derrabus) [inline src/Symfony/Component/VarDumper/Caster/ReflectionCaster.php:364]: 💡 It felt a bit odd that I needed to do this, so I've opened https://bugs.php.net/bug.php?id=80097 for clarification.

**Comment by derrabus** [comment]
The segfault that be observed in https://travis-ci.org/github/symfony/symfony/jobs/726646852 (line 3655) appeared after my changes to the `ReflectionCaster` class. I can reproduce it locally (php 8.0 67d21bf237 on macOS 10.15) by just executing 

```
./phpunit src/Symfony/Component/VarDumper/Tests/Caster/ReflectionCasterTest.php
```

But when I rerun that command a couple of times, it eventually works and the tests pass. I haven't been able to isolate the problem yet, but maybe this is an interesting find for @nikic or @beberlei that helps with stabilizing the reflection API for attributes.

Bildschirmfoto 2020-09-12 um 22 59 17

**Comment by beberlei** [comment]
Do you have the stacktrace for the segfault?

**Review by nicolas-grekas — COMMENTED** [review]
Thanks for the PR.
This should target master to me, as I have this rule in mind:
- making existing code run on newest PHP versions should be considered as a bug fix
- making Symfony compatible with new features of the language should be considered as a new feature.

File: src/Symfony/Component/VarDumper/Tests/Caster/ReflectionCasterTest.php [hunk 2]
```diff
@@ -24,8 +25,12 @@ class ReflectionCasterTest extends TestCase
 {
     use VarDumperTestTrait;
 
-    public function testReflectionCaster()
+    public function testReflectionCasterPhp5()
     {
+        if (\PHP_VERSION_ID >= 70100) {
+            $this->markTestSkipped('This test requires php 7.0 or lower.');
```
[the remark this answers is not in the document]
Comment (derrabus) [reply src/Symfony/Component/VarDumper/Tests/Caster/ReflectionCasterTest.php:31]: I did this first, but it broke the AppVeyor build. Apparently we're running PHPUnit 4.8 on php 7.1 there? 🙀

**Comment by derrabus** [comment]
I use the `dump()` function a lot when debugging on php 8 because I don't have a working PhpStorm/Xdebug-Setup for php 8 yet. It would be extremely helpful to be able to see attributes in the dumped output. But I understand that the policy you mentioned also applies to VarDumper, so I'm going to rebase the PR on master.

File: src/Symfony/Component/VarDumper/Caster/ReflectionCaster.php [hunk 3]
```diff
@@ -105,6 +105,16 @@ public static function castType(\ReflectionType $c, array $a, Stub $stub, bool $
         return $a;
     }
 
+    public static function castAttribute(\ReflectionAttribute $c, array $a, Stub $tsub, bool $isNested)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/VarDumper/Caster/ReflectionCaster.php:108]: s/tsub/stub/

File: src/Symfony/Component/VarDumper/Caster/ReflectionCaster.php [hunk 4]
```diff
@@ -377,7 +405,10 @@ private static function addExtra(array &$a, \Reflector $c)
         }
     }
 
-    private static function addMap(array &$a, \Reflector $c, array $map, string $prefix = Caster::PREFIX_VIRTUAL)
+    /**
+     * @param \Reflector|\ReflectionAttribute $c
```
Comment (nicolas-grekas) [inline src/Symfony/Component/VarDumper/Caster/ReflectionCaster.php:409]: I think we should remove this one: we don't use any method from neither Reflector nor ReflectionAttribute. `object` is just correct here.

**Review by nicolas-grekas — COMMENTED** [review]
with minor comments

**Review by nicolas-grekas — APPROVED** [review]
with minor comments

File: src/Symfony/Component/VarDumper/Caster/ReflectionCaster.php (hunk 3 above)
Comment (derrabus) [reply src/Symfony/Component/VarDumper/Caster/ReflectionCaster.php:108]: 🙈

File: src/Symfony/Component/VarDumper/Caster/ReflectionCaster.php (hunk 4 above)
Comment (derrabus) [reply src/Symfony/Component/VarDumper/Caster/ReflectionCaster.php]: All right, removed.

**Comment by nicolas-grekas** [comment]
for `.github/patch-types.php`:
`src/Symfony/Component/VarDumper/Tests/Fixtures/MyAttribute.php`

**Comment by fabpot** [comment]
Thank you @derrabus.

**Pull request merged**